### PR TITLE
User can save prediction - Adrian Arias

### PR DESCRIPTION
**Pull request by adrianm-arias**
issue-5

### CODE REVIEW

File: client/pages/groups.jsx [hunk 1]
```diff
@@ -30,22 +35,101 @@ export default class Groups extends React.Component {
...
+      body: JSON.stringify(this.state.brackets)
+    })
+      .then(res => res.json())
+      .then(result => {
+        const groupStageCopy = { ...this.state.groupStage };
+        groupStageCopy.bracketId = result.bracketId;
+        this.setState({
+          groupStage: groupStageCopy
+        });
+        fetch('api/groups', {
```
Comment (thebearingedge) [inline client/pages/groups.jsx:96]: If you `return` this `fetch()`, you can delete the `.catch()` on lines 112-114.

File: client/pages/groups.jsx [hunk 2]
```diff
@@ -30,22 +35,101 @@ export default class Groups extends React.Component {
...
   }
 
-  teamSelected(teamId, event) {
+  handleGroupSave() {
+
+    fetch('/api/brackets', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
```
Comment (thebearingedge) [inline client/pages/groups.jsx]: Be sure to send your token along with your request.

File: server/index.js [hunk 3]
```diff
@@ -1,25 +1,130 @@
...
+          const payload = { userId, username, firstName };
+          const token = jwt.sign(payload, process.env.TOKEN_SECRET);
+          res.json({ token, user: payload });
+        });
+    })
+    .catch(err => next(err));
+});
+
+app.post('/api/brackets', (req, res, next) => {
+  const { userId, bracketName } = req.body;
```
Comment (thebearingedge) [inline server/index.js:106]: You will need some authorization middleware. We can't trust the client to include their own `userId` in the request body.

See `api-authorization`

File: server/index.js [hunk 4]
```diff
@@ -1,25 +1,130 @@
...
+      res.status(500).json({
+        error: 'an unexpected error occurred'
+      });
+    });
+});
+
+app.post('/api/groups', (req, res, next) => {
+  const { a1, a2, b1, b2, bracketId, c1, c2, d1, d2, e1, e2, f1, f2, g1, g2, h1, h2 } = req.body;
+  if (!bracketId) {
+    throw new ClientError(400, 'bracket ID are required fields');
```
Comment (thebearingedge) [inline server/index.js:127]: I could be wrong, but I think your userId is needed too, so that you know who saved the group, no?

File: server/index.js [hunk 5]
```diff
@@ -1,25 +1,130 @@
...
+    .then(result => {
+      const [groupSelection] = result.rows;
+      res.status(201).json(groupSelection);
+    })
+    .catch(err => {
+      console.error(err);
+      res.status(500).json({
+        error: 'an unexpected error occurred'
+      });
+    });
```
Comment (thebearingedge) [inline server/index.js]: You can just pass your `err` to `next()` since the error middleware is already set up.

File: server/index.js [hunk 6]
```diff
@@ -1,25 +1,130 @@
...
+    .then(result => {
+      const [brackets] = result.rows;
+      res.status(201).json(brackets);
+    })
+    .catch(err => {
+      console.error(err);
+      res.status(500).json({
+        error: 'an unexpected error occurred'
+      });
+    });
```
Comment (thebearingedge) [inline server/index.js]: You can just pass your `err` to `next()` since the error middleware is already set up.

File: client/pages/home.jsx [hunk 7]
```diff
@@ -1,9 +1,27 @@
...
-    <div>
-      <h1>Home Page</h1>
-    </div>
-  );
+export default class Home extends React.Component {
+
+  render() {
+    const { user } = this.context;
+    // const { firstName } = user;
+    // console.log(user[firstName]);
```
Comment (thebearingedge) [inline client/pages/home.jsx]: Please look through all of your files for commented out code and remove it.

https://github.com/adrianm-arias/final-project/pull/18/files#diff-cc75338894ba5830599852998e0ddb8b3bda389aa0772c5ef479c56842833d14R8-R9

File: client/pages/groups.jsx [hunk 8]
```diff
@@ -228,7 +336,15 @@ export default class Groups extends React.Component {
     }
   }
 
+  componentDidUpdate() {
+    localStorage.setItem('editing-state', JSON.stringify(this.state.isEditing));
+    localStorage.setItem('brackets-state', JSON.stringify(this.state.brackets));
+    localStorage.setItem('groupStage-state', JSON.stringify(this.state.groupStage));
```
Comment (thebearingedge) [inline client/pages/groups.jsx:342]: Sticking this stuff in `localStorage` is a little weird. Why is this necessary?

File: client/components/header.jsx [hunk 9]
```diff
@@ -1,45 +1,59 @@
...
-                <a className={`${route.path === 'login' ? 'nav-link px-3 active' : 'nav-link px-3'}`} href='#login'>
-                  <i className='bi bi-box-arrow-in-right nav-icon-sm' />
-                  Login</a>
+export default class Header extends React.Component {
+
+  render() {
+    const { handleSignOut, route, user } = this.context;
+
+    // const alternateNavLink = !user ? 'Sign In' : 'Sign Out';
+    // const alternateHandle = !user ? 'Sign In' : handleSignOut;
```
Comment (thebearingedge) [inline client/components/header.jsx]: please remove

**Review by thebearingedge — CHANGES_REQUESTED** [review]

File: client/components/header.jsx (hunk 9 above)
Comment (adrianm-arias) [reply client/components/header.jsx]: Done.

File: client/pages/home.jsx (hunk 7 above)
Comment (adrianm-arias) [reply client/pages/home.jsx]: Done.

File: client/pages/groups.jsx (hunk 1 above)
Comment (adrianm-arias) [reply client/pages/groups.jsx:96]: Done.

File: server/index.js (hunk 4 above)
Comment (adrianm-arias) [reply server/index.js:127]: I don't think so, because the group is saved to a bracketId, we set the database this way because there is a second round after group picks the knockout round which will be also linked to the bracketId, that way we know which sets (group and knockout picks)  belong to by the bracket ID, and the bracketID contains a userID to know who that bracket belongs to.

File: client/pages/groups.jsx (hunk 8 above)
Comment (adrianm-arias) [reply client/pages/groups.jsx:342]: After consulting with Scott, I decided to put it in there, since the group stage picks (and the knockout round picks) are a multi step process, to be able to save the users selections incase of internet failure or sudden page refresh, they do not lose their selections and have to restart (they don't need to 'save' or 'send a fetch post request' after every single selection. Instead once they are done with one set (group or knockout selections) then they 'save' and 'send a post request' to send to the database.

File: server/index.js (hunk 6 above)
Comment (adrianm-arias) [reply server/index.js]: Done.

File: server/index.js (hunk 5 above)
Comment (adrianm-arias) [reply server/index.js]: Done.

File: client/pages/groups.jsx (hunk 2 above)
Comment (adrianm-arias) [reply client/pages/groups.jsx]: done.

File: server/index.js (hunk 3 above)
Comment (adrianm-arias) [reply server/index.js:106]: done.

**Review by thebearingedge — APPROVED** [review]
